src/devices/vdml_ext_adi.c

```diff
@@ -0,0 +1,416 @@
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	vexDeviceAdiValueSet(device->device_info, adi_port, value);
+	return_port(smart_port - 1, 1);
+}
+
+int32_t ext_adi_analog_calibrate(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_ANALOG_IN);
+	uint32_t total = 0;
+	for (uint32_t i = 0; i < 512; i++) {
+		total += vexDeviceAdiValueGet(device->device_info, adi_port);
+		task_delay(1); // TODO: If smart ports (and the ADI) only update every 10ms, this really only reads 56 samples, maybe change to a 10ms
+	}
+	adi_data_s_t* const adi_data = &((adi_data_s_t*)(device->pad))[adi_port];
+	adi_data->analog_data.calib = (int32_t)((total + 16) >> 5);
+	return_port(smart_port - 1, (int32_t)((total + 256) >> 9));
+}
+
+int32_t ext_adi_analog_read(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_ANALOG_IN);
+	int32_t rtn = vexDeviceAdiValueGet(device->device_info, adi_port);
+	return_port(smart_port - 1, rtn);
+}
+
+int32_t ext_adi_analog_read_calibrated(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_ANALOG_IN);
+	adi_data_s_t* const adi_data = &((adi_data_s_t*)(device->pad))[adi_port];
+	int32_t rtn = (vexDeviceAdiValueGet(device->device_info, adi_port) - (adi_data->analog_data.calib >> 4));
+	return_port(smart_port - 1, rtn);
+}
+
+int32_t ext_adi_analog_read_calibrated_HR(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_ANALOG_IN);
+	adi_data_s_t* const adi_data = &((adi_data_s_t*)(device->pad))[adi_port];
+	int32_t rtn = ((vexDeviceAdiValueGet(device->device_info, adi_port) << 4) - adi_data->analog_data.calib);
+	return_port(smart_port - 1, rtn);
+}
+
+int32_t ext_adi_digital_read(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_DIGITAL_IN);
+	int32_t rtn = vexDeviceAdiValueGet(device->device_info, adi_port);
+	return_port(smart_port - 1, rtn);
+}
+
+int32_t ext_adi_digital_get_new_press(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	validate_type(device, adi_port, E_ADI_DIGITAL_IN);
+
+	adi_data_s_t* const adi_data = &((adi_data_s_t*)(device->pad))[adi_port];
+
+	int32_t pressed = vexDeviceAdiValueGet(device->device_info, adi_port);
+
+	if (!pressed)
+		adi_data->digital_data.was_pressed = false;
+	else if (!adi_data->digital_data.was_pressed) {
+		// Button is currently pressed and was not detected as being pressed during last check
+		adi_data->digital_data.was_pressed = true;
+		return_port(smart_port - 1, true);
```

Comment:
Pre-existing, but this is some fairly poor debouncing. We might want to improve it, but that's for another pr.

Reply:
if vex doesn't debounce on their side then we posibly end up with a random value anyway, but there's not anything we can do about that really. that said, i think this is mostly useful for getting a latching button